### ISSUE TITLE
Ping/pong response continues after websocket half-close

### ISSUE THREAD

**Issue by drc38** [aaugustin/websockets#1049]
In edge cases, for example when a device is in the process of rebooting, a websocket half-close can happen as per the below:
```
2021-09-15 19:40:00 DEBUG (MainThread) [custom_components.ocpp] Get Configuration for NumberOfConnectors: 1
2021-09-15 19:40:04 DEBUG (MainThread) [websockets.protocol] server > Frame(fin=True, opcode=<Opcode.PING: 9>, data=b'\xb0\x15>\xbd', rsv1=False, rsv2=False, rsv3=False)
2021-09-15 19:40:04 DEBUG (MainThread) [websockets.protocol] server - event = connection_lost([Errno 104] Connection reset by peer)
2021-09-15 19:40:04 DEBUG (MainThread) [websockets.protocol] server - state = CLOSED
2021-09-15 19:40:04 DEBUG (MainThread) [websockets.protocol] server x code = 1006, reason = [no reason]
2021-09-15 19:40:04 DEBUG (MainThread) [websockets.protocol] server - aborted pending pings: 808c3e54, b0153ebd
2021-09-15 19:40:04 DEBUG (MainThread) [websockets.protocol] server ! failing CLOSED WebSocket connection with code 1006
2021-09-15 19:40:04 ERROR (MainThread) [homeassistant] Error doing job: Future exception was never retrieved
websockets.exceptions.ConnectionClosedError: code = 1006 (connection closed abnormally [internal]), no reason
2021-09-15 19:40:04 DEBUG (MainThread) [websockets.protocol] server x half-closing TCP connection
2021-09-15 19:40:04 DEBUG (MainThread) [custom_components.ocpp] Websockets exception: code = 1006 (connection closed abnormally [internal]), no reason
2021-09-15 19:40:04 INFO (MainThread) [custom_components.ocpp] Charger ENX-AU21060165 disconnected from 0.0.0.0:9000.
2021-09-15 19:40:17 DEBUG (MainThread) [websockets.protocol] server > Frame(fin=True, opcode=<Opcode.PING: 9>, data=b'\x87W{\xad', rsv1=False, rsv2=False, rsv3=False)
2021-09-15 19:40:17 DEBUG (MainThread) [websockets.protocol] server - event = data_received(<10 bytes>)
2021-09-15 19:40:17 DEBUG (MainThread) [websockets.protocol] server < Frame(fin=True, opcode=<Opcode.PONG: 10>, data=b'\x87W{\xad', rsv1=False, rsv2=False, rsv3=False)
2021-09-15 19:40:17 DEBUG (MainThread) [websockets.protocol] server - received solicited pong: 87577bad
```
on the server side this is treated as a disconnect, however it is unclear why after the half-close the server continues with ping/pong responses. Is this intended? The client-side then thinks the websocket is still open and continues to send information to the server. The server options are:
```
server = await websockets.serve(
            self.on_connect,
            self.host,
            self.port,
            subprotocols=[self.subprotocol],
            ping_timeout=None,
            close_timeout=10,
        )
```
Appreciate any guidance on the best way to handle the half-close.

**Comment by drc38** [comment]
What I observe sounds similar to issue #622

**Comment by aaugustin** [comment]
This may be fixed in version 10.0, specifically by this line of teh changelog:

**Comment by aaugustin** [comment]
Also, I don't understand how websockets can set a ping on a CLOSED connection (and the connection is in the CLOSED state in your connection) given that `ping` starts with `ensure_open()`. Are you sure you don't have logs from two connections mixed here? Also in websockets 10.0, you can identify which log comes from which connection.

**Comment by drc38** [comment]
Thanks for the quick response, the above log was captured after upgrading to websockets 10.0. There is only 1 websockets server and only 1 client (an ocpp1.6 EV charger) and as far as I can tell the last 4 lines are from the same connection, as there is no new connection established. I have set the websockets logging to debug, is there some way to get more detailed info on the connection?

**Comment by drc38** [comment]
Note at an application level the line `Charger ENX-AU21060165 disconnected from 0.0.0.0:9000` is generated from the finally statement in the on_connect function.

**Comment by drc38** [comment]
> Thanks for the quick response, the above log was captured after upgrading to websockets 10.0. There is only 1 websockets server and only 1 client (an ocpp1.6 EV charger) and as far as I can tell the last 4 lines are from the same connection, as there is no new connection established. I have set the websockets logging to debug, is there some way to get more detailed info on the connection eg unique identifier?

Sorry looks like I was using 9.1 for that log, this is a sample log for 10.0, how do I identify connection from the log?
```
2021-09-17 10:04:38 DEBUG (MainThread) [websockets.server] %% sending keepalive ping
2021-09-17 10:04:38 DEBUG (MainThread) [websockets.server] > PING 1c ab 45 1d [binary, 4 bytes]
2021-09-17 10:04:39 DEBUG (MainThread) [websockets.server] < PONG 1c ab 45 1d [binary, 4 bytes]
```

**Comment by drc38** [comment]
> This may be fixed in version 10.0, specifically by this line of the changelog:
> 
> * Avoided a crash when receiving a ping while the connection is closing.
> 
> Are you able to test on 10.0 and see if it happens again?
> 
> EDIT - I was confused; we're sending pings here, not receiving them.

Does appear to be fixed in 10.0, thanks!

**Comment by aaugustin** [comment]
Good to hear it's fixed – apparently the cleanup I did in this area was needed.

Adding the connection id to logs requires a bit of configuration, see here: https://websockets.readthedocs.io/en/stable/topics/logging.html#configure-logging
